pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java

```diff
@@ -18,85 +18,148 @@
  */
 package org.apache.pinot.core.data;
 
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import java.io.IOException;
-import java.util.Arrays;
-import java.util.HashMap;
 import java.util.Map;
 import java.util.Set;
-import org.apache.pinot.common.data.RowEvent;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.utils.EqualityUtils;
 import org.apache.pinot.common.utils.JsonUtils;
+import org.apache.pinot.core.data.readers.RecordReader;
+import org.apache.pinot.core.data.recordtransformer.NullValueTransformer;
+import org.apache.pinot.core.data.recordtransformer.RecordTransformer;
 
 
 /**
- * A plain implementation of RowEvent based on HashMap. Should be reused as much as possible via
- * {@link GenericRow#createOrReuseRow(GenericRow)}
+ * The generic row is the value holder returned from {@link RecordReader#next()} and
+ * {@link RecordReader#next(GenericRow)}, and can be modified with {@link RecordTransformer}. The generic row returned
+ * from the {@link NullValueTransformer} should have {@code defaultNullValue} filled to the fields with {@code null}
+ * value, so that for fields with {@code null} value, {@link #getValue(String)} will return the {@code defaultNullValue}
+ * and {@link #isNullValue(String)} will return {@code true}.
  */
-public class GenericRow implements RowEvent {
-  private Map<String, Object> _fieldMap = new HashMap<>();
+public class GenericRow {
+  private final Map<String, Object> _fieldToValueMap = new Object2ObjectOpenHashMap<>();
```

Comment:
is fastutil really needed? at some point we want to move GenericRow and RecordReader out of pinot core and part of connectors (batch and real-time) into a separate module that external users can use. Its better to keep the dependency to a minimum.

Reply:
Make sense. Fastutil map is faster, but maybe not necessary. Will replace with java map.